Boundary Docs: Configure workers refactor (#3077)
hashicorp/web-unified-docs: content/boundary/v1.0.x/content/docs/monitor/health.mdx

### Worker health requirements

To enable the worker health endpoint, define a `purpose = "ops"` listener in the worker's own configuration file.
The worker serves its own health endpoint, so you do not query the controller to determine the health of a worker.

If a single Boundary instance runs both a controller and a worker, the instance serves the controller health endpoint.
In that case, you request the worker information from the controller endpoint using the `worker_info` parameter.

## Shutdown grace period

When the controller health endpoint is enabled, you can configure the controller to respond with `503 Service Unavailable` when it receives a shutdown signal.
The controller then waits a configurable amount of time before it starts the shutdown process.

This feature supports load balancing by reducing the risk of an outgoing Boundary instance causing disruption to incoming requests.

In this state, Boundary still processes requests as normal, but reports as unhealthy through the controller health endpoint.
In load-balanced environments, the load balancer removes the unhealthy instance from the pool of instances eligible to handle requests.

> This reduces the likelihood that the instance receives a request during shutdown.

Reviewer: ```suggestion
Removing the instance reduces the likelihood that it receives a request during shutdown.
```

| `5xx`        | `GET /health` returns HTTP status `5XX` or request timeout if unhealthy |
| `503`        | `GET /health` returns HTTP status `503 Service Unavailable` status if the controller is shutting down |

All controller responses return empty bodies. `GET /health` does not support any input.

The worker health service uses the same path, but it accepts the `worker_info` query parameter and returns its status in the response body:

| Request | Response |
|---------|----------|
| `GET /health` | Returns `200 OK` with an empty JSON object, `{}` |
| `GET /health?worker_info=1` | Returns `200 OK` with a `worker_process_info` object |

<Warning>

  The worker health endpoint always returns `200 OK`, even when the worker cannot reach its upstream.
  A health check that tests only the HTTP status code reports a disconnected worker as healthy.

  To determine whether a worker is ready to proxy sessions, request the worker information and check the `upstream_connection_state` field.

> </Warning>

Reviewer: ```suggestion
<Note>

  The worker health endpoint always returns `200 OK`, even when the worker cannot reach its upstream.
  A health check that tests only the HTTP status code reports a disconnected worker as healthy.

  To determine whether a worker is ready to proxy sessions, request the worker information and check the `upstream_connection_state` field.

</Note>
```

Reviewer: We haven't always used alerts correctly, and the skill might be picking up on some of our bad examples. Per our standards:

- Warning should be information the user must take action on. Should only be used for breaking changes and security vulnerabilities.
- Note should be information that the user may need to take action on.
- Tip should be a best practice or optional setting/workflow. They are not required for users to complete a procedure.

I think this one should be a note.

## Check the health endpoint using `wget`

The Boundary Docker image includes `wget`. You can use it to check the health endpoint. Enterprise edition users can check the health of controllers and workers. HCP Boundary users can check the health of their self-managed workers.

Use the following command to check whether a worker is connected to its upstream:

```shell-session
$ wget -q -O - http://localhost:9203/health?worker_info=1 | grep -q 'READY'
```

- `-q` - Runs in quiet mode, which suppresses the progress output that `wget` writes to `stderr`.
- `-O -` - Writes the response body to `stdout` instead of saving it to a file.

The `grep -q 'READY'` command exits with code `0` if the worker's upstream connection is in the `READY` state, and a non-zero code otherwise.

> Because the worker endpoint always returns `200 OK`, you must inspect the response body like this rather than rely on the HTTP status code.

Reviewer: ```suggestion
Because the worker endpoint always returns `200 OK`, you must inspect the response body rather than rely on the HTTP status code.
```

- `upstream_connection_state` - The connection state of the worker. This value indicates whether the worker can connect to an upstream address or connection. It can be any of the following states:
   - `CONNECTING` - The channel is trying to make a connection and is waiting for name resolution or the connection establishment.
   - `READY` - The channel has successfully established a connection, and any attempts to communicate have succeeded.
   - `TRANSIENT_FAILURE` - The channel suffered a transient failure such as a time out or socket error.
   A channel in this state switches to the `CONNECTING` state and tries to establish a connection.
   - `IDLE` - The channel is not attempting to create a connection because there are no calls.
   - `SHUTDOWN` - The channel is shutting down.
   Any new calls fail immediately.
   Pending calls may continue running until Boundary cancels them.

<Note>

  Boundary omits fields that are set to their zero value from the response.
  A worker that has never connected to an upstream may return only the fields that have values.

> </Note>

Reviewer: ```suggestion
Boundary omits fields that are set to their zero value from the response.
 A worker that has never connected to an upstream may return only the fields that have values.
```

Reviewer: Since this is just informative, I don't think it should be a note.
